torch/testing/_asserts.py

```diff
@@ -102,6 +97,63 @@ def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Excepti
     return wrapper
 
 
+def _check_quantization(check_tensor_values: Callable[..., Optional[Exception]]) -> Callable[..., Optional[Exception]]:
+    """Decorates continuous tensor check functions to handle quantized tensors.
+
+    If the inputs are not quantized, this decorator is a no-op.
+
+    Args:
+        check_tensor_values (Callable[..., Optional[Exception]]): Tensor check function for continuous tensors.
+
+    Returns:
+        Optional[Exception]: Return value of :attr:`check_tensors`.
+    """
+
+    @functools.wraps(check_tensor_values)
+    def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Exception]:
+        if not actual.is_quantized:
+            return check_tensor_values(actual, expected, **kwargs)
+
+        if actual.qscheme() == torch.per_tensor_affine:
+            exc = check_tensor_values(torch.as_tensor(actual.q_scale()), torch.as_tensor(expected.q_scale()), **kwargs)
+            if exc:
+                return _amend_error_message(exc, "{}\n\nThe failure occurred for the quantization scale.")
+
+            exc = check_tensor_values(
+                torch.as_tensor(actual.q_zero_point()), torch.as_tensor(expected.q_zero_point()), **kwargs
+            )
+            if exc:
+                return _amend_error_message(exc, "{}\n\nThe failure occurred for the quantization zero point.")
+        elif actual.qscheme() == torch.per_channel_affine:
+            exc = check_tensor_values(actual.q_per_channel_scales(), expected.q_per_channel_scales(), **kwargs)
+            if exc:
+                return _amend_error_message(exc, "{}\n\nThe failure occurred for the quantization scales.")
+
+            exc = check_tensor_values(
+                actual.q_per_channel_zero_points(), expected.q_per_channel_zero_points(), **kwargs
+            )
+            if exc:
+                return _amend_error_message(exc, "{}\n\nThe failure occurred for the quantization zero points.")
+
+            exc = check_tensor_values(
+                torch.as_tensor(actual.q_per_channel_axis()), torch.as_tensor(expected.q_per_channel_axis()), **kwargs
+            )
+            if exc:
+                return _amend_error_message(exc, "{}\n\nThe failure occurred for the quantization per-channel axis.")
+        elif actual.qscheme() == torch.per_tensor_symmetric:
+            # TODO add meta data checking
+            pass
+        elif actual.qscheme() == torch.per_channel_symmetric:
+            # TODO add meta data checking
+            pass
+        else:
+            raise UsageError(f"Unknown quantization scheme {actual.qscheme()}.")
+
+        return check_tensor_values(actual.int_repr().to(torch.int32), expected.int_repr().to(torch.int32), **kwargs)
```

Comment:
Is there an advantage from `.int_repr().to(torch.int32)` over `.dequantize()`. IMO the latter implies the intent better.

Reply:
int_repr() is comparing integer representations of a quantized tensor and dequantize will compare floating point values (which might have some tolerance), I think int_repr will be a more accurate comparison

But dequantize() is more general I think, it would work for all types of quantized tensors, it's OK to use that as well since we also check other quantization parameters as well

Reply:
Note that in general, `torch.testing.assert_close` is a closeness rather than an equality check. Without passing an `rtol` or `atol` they will default to `0` since a quantized tensor has an integral dtype. From this perspective I have two questions:

1. Should any of the other members `q_scale()`, `q_zero_point()` checked for closeness rather than equality?
2. Especially for closeness checks, `dequantize()` makes more sense, correct? Otherwise the tolerances do not relate to the actual values.

Reply:
I see, I thought this is checking for equality. Maybe for closeness we do not really need to check for the quantization parameters? I feel closeness of two quantized Tensor should be defined as closeness of the dequantized Tensor, what do you think?

Reply:
I'm sorry, I think I got the meaning of `.int_repr()` and `dequantize()` wrong. After some experimentation, this is my current understanding: `.int_repr()` actually computes `round(x / s + z)` whereas `.dequantize()` will give you `x`, albeit with quantization errors. 

Is that correct? If yes `.dequantize()` should be treated similar to `.q_scale()` and `.q_zero_point()` since it checks part of the equation rather than the result. 

---

> Maybe for closeness we do not really need to check for the quantization parameters? I feel closeness of two quantized Tensor should be defined as closeness of the dequantized Tensor, what do you think?

Note that `torch.testing.assert_close` is not just a pure closeness check like `torch.isclose`, but is supposed to be used in a test setting where you want to for example test that an operator behaves the way it should. Consider the following example where `actual` was generated by said operator: 

```python
import torch

actual = torch.quantize_per_tensor(
    torch.tensor([-1.0, 3.0]),
    scale=0.5,
    zero_point=-2,
    dtype=torch.qint32,
)
expected = torch.quantize_per_tensor(
    torch.tensor([12.0, -4.0]),
    scale=-2,
    zero_point=2,
    dtype=torch.qint32,
)
```

Would you consider this correct behavior of the operator? Because the `.int_repr()` of the tensors is equal:

```python
>>> print(actual.int_repr())
tensor([-4,  4], dtype=torch.int32)
>>> print(expected.int_repr())
tensor([-4,  4], dtype=torch.int32)
```

Thus, if we have access to all parts of the quantization equation, we should test each for closeness / equality. Since the scale and zero point is stored without errors, we should check these for equality, whereas `.dequantize()` can be checked for closeness (not by default though, because the tolerances for integral dtypes default to `0`).

Reply:
int_repr is the integer value of the quantized Tensor, dequantize() returns the value of quantized Tensor in floating point. dequantize() return a Tensor that is represented by the quantized Tensor, closeness of the different parts of quantized Tensor does not directly translate to closeness of the floating point value I think and I feel it's not really meaningful to compare the parts of quantized Tensor since the value of quantized Tensor is defined by all components (scale, zero_point, int_repr), I feel individual components doesn't have a meaning by itself.

and tolerance should only be defined in the floating point relam as well I think.

Reply:
Any thoughts on adding back the checks for scale and zero_point equality here?  A motivating example could be a tensor where all of the values are 0, the zero_point is 0, and the scale could be any number without changing the value of the dequantized version of the tensor.  If we have a test case which only checks the dequantized version, it would not catch any changes in scale.  I personally would appreciate the extra safety of knowing that any change in quantized tensor internals would be flagged by these checks, even if the change does not change the dequantized output.

Reply:
I've created #68548 so we don't have to discuss it here on this already merged PR.